### PR TITLE
nixos/systemd-stage-1: Support resolved

### PR DESCRIPTION
**Pull request by ElvishJerricco**
## Description of changes

This comes with an additional change to put `systemd-resolved.service` in `sysinit.target` instead of `multi-user.target`, given that this is what the upstream unit file indicates in its `[Install]` section.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by ElvishJerricco** [comment]
This still needs a NixOS test. Help would be greatly appreciated on this, because I'm not actually all that familiar with how to deal with DNS and have already spent hours trying to write a test that I'm pretty sure is nowhere near correct.

**Comment by ElvishJerricco** [comment]
So, I've added a test for `systemd-resolved`, but I'm not *quite* satisfied with it in stage 1. I couldn't do the `resolvectl query` part of the test in stage 1 for some reason. I get this:

```
example.com: resolve call failed: The name org.freedesktop.resolve1 was not provided by any .service files
```

And looking back at the logs, I sometimes see something like this,

```
client # [    1.333575] systemd[1]: systemd-networkd.service: Unexpected error response from GetNameOwner(): Connection terminated
client # [    1.334477] systemd-sysroot-fstab-check[149]: Requesting swap.target/start/replace...
client # [    1.341035] systemd[1]: systemd-resolved.service: Unexpected error response from GetNameOwner(): Connection terminated
```

Or sometimes something like this,

```
client # [    1.311828] systemd[1]: Failed to subscribe to NameOwnerChanged signal for 'org.freedesktop.network1': Transport endpoint is not connected
client # [    1.312865] systemd[1]: systemd-networkd.service: Cannot watch bus name org.freedesktop.network1: Transport endpoint is not connected
client # [    1.316867] systemd[1]: Failed to subscribe to NameOwnerChanged signal for 'org.freedesktop.resolve1': Transport endpoint is not connected
client # [    1.317882] systemd[1]: systemd-resolved.service: Cannot watch bus name org.freedesktop.resolve1: Transport endpoint is not connected
```

Or sometimes nothing indicative of any problems. No matter the logs, I get the `resolve call failed` error from `resolvectl`.

**Comment by arianvp** [comment]
Sounds like dbus is not running

**Comment by ElvishJerricco** [comment]
@arianvp It is though. If I add `print(client.succeed("systemctl status dbus.socket dbus.service"))`, I get this:

```
client: must succeed: systemctl status dbus.socket dbus.service
(finished: must succeed: systemctl status dbus.socket dbus.service, in 0.01 seconds)
● dbus.socket - D-Bus System Message Bus Socket
     Loaded: loaded (/etc/systemd/system/dbus.socket; linked; preset: enabled)
     Active: active (running) since Tue 2024-05-21 07:26:24 UTC; 5s ago
   Triggers: ● dbus.service
     Listen: /run/dbus/system_bus_socket (Stream)
     CGroup: /system.slice/dbus.socket

May 21 07:26:24 localhost systemd[1]: Listening on D-Bus System Message Bus Socket.

● dbus.service - D-Bus System Message Bus
     Loaded: loaded (/etc/systemd/system/dbus.service; linked; preset: enabled)
     Active: active (running) since Tue 2024-05-21 07:26:24 UTC; 5s ago
TriggeredBy: ● dbus.socket
       Docs: man:dbus-daemon(1)
   Main PID: 99 (dbus-daemon)
      Tasks: 1 (limit: 1131)
     Memory: 388.0K (peak: 652.0K)
        CPU: 7ms
     CGroup: /system.slice/dbus.service
             └─99 /nix/store/pvyfwb5fyh0max87rkhch4ac4d9b90lq-dbus-1.14.10/bin/dbus-daemon --system --address=systemd: --nofork --nopidfile --systemd-activation --syslog-only

May 21 07:26:24 localhost systemd[1]: Starting D-Bus System Message Bus...
May 21 07:26:24 localhost systemd[1]: Started D-Bus System Message Bus.
```

**Comment by ElvishJerricco** [comment]
Ah, we didn't have the necessary dbus files from systemd included. Fixed, and ready for review!

File: nixos/modules/system/boot/resolved.nix [hunk 1]
```diff
@@ -126,60 +140,87 @@ in
...
+
+        # symlink the dynamic stub resolver of resolv.conf as recommended by upstream:
+        # https://www.freedesktop.org/software/systemd/man/systemd-resolved.html#/etc/resolv.conf
+        "resolv.conf".source = "/run/systemd/resolve/stub-resolv.conf";
+      } // optionalAttrs dnsmasqResolve {
+        "dnsmasq-resolv.conf".source = "/run/systemd/resolve/resolv.conf";
+      };
+
+      # If networkmanager is enabled, ask it to interface with resolved.
+      networking.networkmanager.dns = "systemd-resolved";
```
Comment (flokli) [inline nixos/modules/system/boot/resolved.nix:191]: networkmanager in initrd too? ;-)

Comment (flokli) [reply nixos/modules/system/boot/resolved.nix:191]: But then let's remove that line for now, this is copypasta from elsewhere and might cause confusion.

Comment (ElvishJerricco) [reply nixos/modules/system/boot/resolved.nix:191]: @flokli I'm confused. That line is not in this PR. It's a pre-existing line and it's there to configure networkmanager in stage 2. It's not for this at all.

Comment (ElvishJerricco) [reply nixos/modules/system/boot/resolved.nix:191]: @flokli oh I see. I had "Hide whitespace" on in github. That part of the diff is from the re-indent commit. It's just a whitespace change. The line is there for stage 2 purposes and is unchanged.

**Review by arianvp — APPROVED** [review]

**Comment by philiptaron** [comment]
This is really sweet! Thank you for your continued grind on making systemd in stage1 work, @ElvishJerricco. Appreciate you.

**Pull request merged**